### DISCUSSION TITLE
Airplay2 - wireless speakers - stereo - latency ?

### DISCUSSION THREAD

**Discussion by klslz** [mikebrady/shairport-sync#1461] [General]
Hi. 

I was wondering if shairport (AP2) could be used to build a system of two wireless active speakers running 
an RPi on either side.
There are commercial solutions for wireless systems (e.g. Dali). Somehow they managed to get a wireless system properly in sync. 
There has to be an advanced latency control and buffer management  in place to accomplish a good working "sync" setup between the left and the right speaker

AP2 streamers like the Wiim Mini Airplay 2 Music Streamer seem to use microphones to (auto-)sync clients properly. Not sure how well that works.

I'd guess the key challenge is not just keep both RPis in whatever "sync" though. 
If both systems should act as a high quality stereo config, we probably talk about "single-digit-milliseconds" left-right sync and a very low level of interchannel variations (data jitter). 

Anybody tried this with shairport !?!? Is it possible? Would it be possible at all?

**Comment by npmstart-pray** [comment]
Balena-sound is what you're looking for.

**Reply by klslz** [reply]
I don't think so.

**Reply by charlesomer** [reply]
I guess you could connect a Pi Zero W v2 to each speaker and run Shairport Sync which would give you the ability to play to both of them together in sync. However since it sounds like you want to use them similar to how HomePod's work in stereo pairs? I don't think it's currently possible to setup SPS as a stereo pair so you'd probably be relying on the source (whatever that might be) to connect to both of them and send them the audio you want (left and right channels in your case).

**Comment by klslz** [comment]
shairport-sync , afaik, was built having lowest latencies between synced clients in mind. 
From what I understood airplay2 buffers large parts of the audio track on a client. It then gets playbacked in a controlled and synced fashion. But it's one thing to keep 2 rooms in sync. Small non-linearities/out-of-sync issues wouldn't matter.

The basic question is. Is a proper shairport-sync installation/configuration sufficient to run separate stereo speakers without serious interchannel non-linearities?  At a quality level rivalling classic stereo systems.

**Reply by charlesomer** [reply]
From [AIRPLAY2.md](https://github.com/mikebrady/shairport-sync/blob/development/AIRPLAY2.md)

> Two types of audio are received by Shairport Sync -- "Realtime" streams of CD quality ALAC (like AirPlay 1) and "Buffered Audio" streams of AAC stereo at 44,100 frames per second. The selection of stream type is made by the player.

Realtime streams should have the audio playing at the correct time as the content on screen.

There's some information about sync accuracy and the ability to adjust it here: [resynchronisation](https://github.com/mikebrady/shairport-sync/blob/development/MOREINFO.md#resynchronisation) & [tolerance](https://github.com/mikebrady/shairport-sync/blob/development/MOREINFO.md#tolerance)

**Reply by mikebrady** [reply]
> The basic question is. Is a proper shairport-sync installation/configuration sufficient to run separate stereo speakers without serious interchannel non-linearities? At a quality level rivalling classic stereo systems.

Unfortunately I think not. The timing inaccuracies in Shairport Sync must be of the order of milliseconds or hundreds of microseconds.  Apple can do this to a higher degree of accuracy, apparently, but I have no real idea how.

**Comment by noelhibbard** [comment]
This setting in the config file should accomplish what you are after:
```playback_mode = "stereo"; // This can be "stereo", "mono", "reverse stereo", "both left" or "both right". Default is "stereo".```

Set one RPi to "both left" and the other RPi to "both right".

**Comment by klslz** [comment]
Sorry, for not being clear.

For quality stereo playback the time difference between left and right channel would be the key challenge here.
E.g. 1 millisecond difference between left and right would equal a 34cm distance  >>  one speaker 34cm closer to you.
That means we'd need to look at a double digit micro second demand on that front  >> 1cm = 30us.

On a single standalone system this is no issue. We can configure delays via software. On two seperate systems with drifting clocks, different setups etc, that would be a major challenge. We simply can't assume both client systems work equally timewise. There'll be variations. To cope with that is most challenging.

Therefore it IMO would require a sophisticated buffered sync mechanism between left and right client on a lower microsecond level to accomplish quality stereo playback. Just by looking at typical double digit milliseconds latencies on non-realtime systems this to me looks like a major challenge.

Anyhow. Perhaps there's a solution to it. Mike and team did put a lot of effort on the latency and sync subject.

@noelhibbard 
If we'd run left/right channel stereo on two separate clients I'd prefer to have a "single left" and "single right" mode for each channel to reduce the data transfer by 50%.

**Reply by noelhibbard** [reply]
@klslz AirPlay is only stereo so regardless of what shairplay-sync does, the iDevice will always stream 2 channels to both endpoints. This setting just tells shairport-sync to drop one of the input channels and use the other for both output channels. This feature was added at the request of another user who was attempting the same thing you are. If you searched the issues you may be able to find this person and reach out to see if they had success with the latency issue.

**Comment by Hk1020** [comment]
I am entertaining a similar but less sophisticated idea. I own two battery driven Canton musicboxes xs which have excellent sound but are AUX and bluetooth only. They do stereo via bluetooth true wireless. I managed to hook one of them up to a Pi 3b with an aux cable (so far only to the headphone jack) and they play nicely as a stereo pair that way.

I am thinking of using a Pi zero 2 W to make this somewhat portable. Would a Pi zero actually be powerful enough for Airplay 2? And I also don’t have a good idea about how to power the Pi. There is nothing on the speaker I could tap into.  These boxes also add a lot of latency which I could compensate with `audio_backend_latency_offset_in_seconds` in the config. Is there a better way than trial and error to determine the delay?

**Comment by mikebrady** [comment]
Thanks for the post.

* What has emerged is that the Pi Zero W just fails to be powerful enough. In a “steady state” on a good network with accurate timing, it just about works. But if a lot of interpolation is needed or if you switch to a new track, for example, it can drop out for a few seconds. The Pi Zero 2 W is fine.
* Trial and error, I’m afraid. Playing a video of a clapper board can help.

**Comment by Hk1020** [comment]
I'd like to get back to this as today I got around to really fix the delay by measuring it.

I recorded an audio track by tapping on the iPad with a pen in semi-regular intervals of about 6 seconds. This track I played simultaneously on a regular airplay speaker and via shairport on the Canton speakers. The resulting very audible double clicks I recorded again and analyzed the peaks in audacity on a PC. A script I found [here](https://forum.audacityteam.org/viewtopic.php?p=278483#p278483) gave me very accurate results of the times when the peaks started. Computing the differences (0.3628 sec) gave me the delay. The result is so accurate that holding the left channel from the regular box to one ear and the right channel of the Canton boxes to the other gives perfect sound (disregarding the speaker differences). I am thrilled.

But I have a question. A small problem is that the delay is very different when I use a single box, so I need a different config. Do you have a good suggestion how to switch between the two setups? For example, could you run two shairports on the same computer simultaneously? But it is a minor problem, the single box doesn't sound very good anyway.

**Comment by mikebrady** [comment]
Thanks for the update.

> Could you run two shairports on the same computer simultaneously?

With difficulty -- it seems that AirPlay 2 clients can not deal with more than one player on a single device, so you have to do some virtualisation or containerisation, which brings its own complications. @noelhibbard is the expert here.
